### PR TITLE
ROX-15458, ROX-16687: Add resending NodeInventory when no ACK arrives

### PR DESCRIPTION
**Pull request by vikin91**
## Description

This PR changes the behavior of Compliance with respect to [N]ACK messages for the Node Inventories.
On sending node inventory, each compliance container will now:
- start a timer,
- clone the node inventory and store it in memory

If there is no ACK within time defined by `ROX_NODE_SCANNING_ACK_DEADLINE_BASE`, then the node inventory will be resent from the copy held in cache. Before each N-th attempt to resent, compliance will wait `N * ROX_NODE_SCANNING_ACK_DEADLINE_BASE`. If the ACK arrives, the cache is nilled and the ticker is stopped.
When there is no ACK but a new node-inventory is scheduled (i.e., after 4 hours), the old cache is dropped, the ticker is
reset and the whole process starts from scratch. At any point in time, there is maximally one node-inventory object held in cache in compliance container.

This PR also removes the NACK mechanism between Sensor and Compliance, as it is completely replaced by the lack of ACK. If Sensor emits NACK, then Compliance will behave as if there is no ACK = retry.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added

### N/A
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

## Testing Performed

- Manual deployment and testing the negative and positive scenarios using `local-compliance`:
    - Positive: ACK arrives to Compliance and the inventory is not resent
    - Negative: ACK does not arrive to Compliance and the signal to resent the inventory is emitted
- Manual tests as above on multiple secured clusters to make sure that the routing of ack messages works correctly

#### ACK arrives before the first retry

```
│ pkg/logging: 2023/06/01 10:57:44.852514 logger.go:170: Info: Scanning node "piotr-work-30may-npst5-worker-b-62rlb"                                                                  │
│ pkg/logging: 2023/06/01 10:57:44.859013 logger.go:170: Info: Observing message scan being sent. Waiting for an ACK for 15s                                                          │
│ pkg/logging: 2023/06/01 10:57:44.859130 logger.go:170: Info: Next node scan in 41.425088147s                                                                                        │
│ pkg/logging: 2023/06/01 10:57:47.526111 logger.go:165: Info: Message has been acknowledged
```

#### No ACK arrives 

`ROX_NODE_SCANNING_ACK_DEADLINE_BASE=5s`

```
pkg/logging: 2023/06/05 15:55:37.364540 logger.go:170: Info: Scanning node "local-compliance"
pkg/logging: 2023/06/05 15:55:37.364615 logger.go:170: Info: Generating Node Inventory
pkg/logging: 2023/06/05 15:55:37.364632 logger.go:170: Info: Next node scan in 1m0s
pkg/logging: 2023/06/05 15:55:42.364865 logger.go:170: Info: Suggesting to resend, retry 1 (next retry in 5s)
pkg/logging: 2023/06/05 15:55:47.365982 logger.go:170: Info: Suggesting to resend, retry 2 (next retry in 10s)
pkg/logging: 2023/06/05 15:55:57.366706 logger.go:170: Info: Suggesting to resend, retry 3 (next retry in 15s)
pkg/logging: 2023/06/05 15:56:12.367345 logger.go:170: Info: Suggesting to resend, retry 4 (next retry in 20s)
pkg/logging: 2023/06/05 15:56:32.368884 logger.go:170: Info: Suggesting to resend, retry 5 (next retry in 25s)
pkg/logging: 2023/06/05 15:56:37.366601 logger.go:170: Info: Scanning node "local-compliance"
pkg/logging: 2023/06/05 15:56:37.366662 logger.go:170: Info: Generating Node Inventory
pkg/logging: 2023/06/05 15:56:37.366756 logger.go:170: Info: Next node scan in 1m0s
pkg/logging: 2023/06/05 15:56:42.367714 logger.go:170: Info: Suggesting to resend, retry 1 (next retry in 5s)
pkg/logging: 2023/06/05 15:56:47.367896 logger.go:170: Info: Suggesting to resend, retry 2 (next retry in 10s)
```

### CODE REVIEW

File: pkg/retry/unconfirmed_message_handler.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+// ObserveSending should be called when a new message is sent and it is expected to be [N]ACKed
+func (s *UnconfirmedMessageHandlerImpl) ObserveSending() {
+	s.retryMux.Lock()
+	defer s.retryMux.Unlock()
+	log.Debugf("Observing message being sent. Waiting for an ACK for %s", s.baseInterval.String())
+	s.ticker.Stop()
+	s.retry = 0
+	s.ticker.Reset(s.baseInterval)
+}
```
Comment (vikin91) [inline pkg/retry/unconfirmed_message_handler.go:97]: On 2nd an later exec of this function, we should not reset the ticker.

File: pkg/retry/unconfirmed_message_handler.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	// resendInterval is a multiply of the baseInterval and defines how much time should pass before the next retry
+	resendInterval time.Duration
+	// ticker controls the delay after which we resend a message
+	ticker *time.Ticker
+	// ch will produce a message when the message should be resent
+	ch chan struct{}
+	// retry counts the number of retries for a given message
+	retry int
+	// retryMux is a mutex for retry variable
+	retryMux *sync.Mutex
```
Comment (vikin91) [inline pkg/retry/unconfirmed_message_handler.go:31]: `retry` could be an atomic integer

Comment (vikin91) [reply pkg/retry/unconfirmed_message_handler.go:31]: I tried to change `s.retry++` to `atomic.AddInt32(&s.retry, int32(1))` and removing the mutex, but this causes race to detected.

**Review by fredrb — APPROVED** [review]

File: pkg/retry/unconfirmed_message_handler.go (hunk 2 above)
Comment (vikin91) [reply pkg/retry/unconfirmed_message_handler.go:31]: After the recent changes, the atomic + retry actually does the job. The problem is now with `if s.numUnackedSendings > 1 {`

File: pkg/retry/unconfirmed_message_handler.go [hunk 3]
```diff
@@ -90,26 +84,22 @@ func (s *UnconfirmedMessageHandlerImpl) retryLater() {
 
 // ObserveSending should be called when a new message is sent and it is expected to be [N]ACKed
 func (s *UnconfirmedMessageHandlerImpl) ObserveSending() {
-	s.retryMux.Lock()
-	defer s.retryMux.Unlock()
-	s.numUnackedSendings++
+	s.numUnackedSendings.Add(1)
 	log.Debugf("Observing message being sent. Waiting for an ACK for %s", s.baseInterval.String())
-	if s.numUnackedSendings > 1 {
+	if !s.numUnackedSendings.CompareAndSwap(1, 1) { // if s.numUnackedSendings > 1
```
Comment (fredrb) [inline pkg/retry/unconfirmed_message_handler.go:89]: Why use CAS? Wouldn't 
```
if s.numUnackedSendings.Load() > 1 {
```
Yield the same results?

Using CAS here, you're effectively checking if the number is `1` and if it is updating it to `1`.

Comment (vikin91) [reply pkg/retry/unconfirmed_message_handler.go:89]: Yeah, that works, thanks!

**Review by fredrb — APPROVED** [review]
Let's 🚢

**Review by lvalerom — APPROVED** [review]
Looks good! 🚀

**Comment by vikin91** [comment]
Failure on `ocp-4-13-qa-e2e-tests` are due to collector crashing being unable to download kernel objects from sensor. [Slack thread checking with Collector team](https://redhat-internal.slack.com/archives/CFMQ5C2TT/p1686133904148159)

**Comment by vikin91** [comment]
Failure in `cp-4-12-core-bpf-qa-e2e-tests` is addressed by [ROX-17557](https://issues.redhat.com/browse/ROX-17557)

**Pull request merged**